Bugsnag/BugsnagSystemState.m

```diff
@@ -180,47 +185,71 @@ - (instancetype)initWithConfiguration:(BugsnagConfiguration *)config {
+            [weakSelf setValue:@NO forAppKey:SYSTEMSTATE_APP_IS_IN_FOREGROUND];
         }];
         [center addObserverForName:UIApplicationDidBecomeActiveNotification object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
             [weakSelf.kvStore setBoolean:YES forKey:SYSTEMSTATE_APP_IS_ACTIVE];
-            [weakSelf bgSetAppValue:@YES forKey:SYSTEMSTATE_APP_IS_ACTIVE];
+            [weakSelf setValue:@YES forAppKey:SYSTEMSTATE_APP_IS_ACTIVE];
         }];
         [center addObserverForName:UIApplicationWillResignActiveNotification object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
             [weakSelf.kvStore setBoolean:NO forKey:SYSTEMSTATE_APP_IS_ACTIVE];
-            [weakSelf bgSetAppValue:@NO forKey:SYSTEMSTATE_APP_IS_ACTIVE];
+            [weakSelf setValue:@NO forAppKey:SYSTEMSTATE_APP_IS_ACTIVE];
         }];
         [center addObserverForName:UIApplicationDidReceiveMemoryWarningNotification object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
             NSString *date = [BSG_RFC3339DateTool stringFromDate:[NSDate date]];
             [weakSelf.kvStore setString:date forKey:SYSTEMSTATE_APP_LAST_LOW_MEMORY_WARNING];
-            [weakSelf bgSetAppValue:date forKey:SYSTEMSTATE_APP_LAST_LOW_MEMORY_WARNING];
+            [weakSelf setValue:date forAppKey:SYSTEMSTATE_APP_LAST_LOW_MEMORY_WARNING];
         }];
 #endif
     }
     return self;
 }
 
+- (void)recordAppUUID {
+    // [BSG_KSSystemInfo appUUID] returns nil until we have called _dyld_register_func_for_add_image()
+    [self setValue:[BSG_KSSystemInfo appUUID] forAppKey:BSGKeyMachoUUID];
+}
+
+- (void)setBatteryCharging:(BOOL)charging {
+    [self setValue:@(charging) forDeviceKey:BSGKeyCharging];
+}
+
+- (void)setBatteryLevel:(NSNumber *)batteryLevel {
+    [self setValue:batteryLevel forDeviceKey:BSGKeyBatteryLevel];
+}
+
 - (void)setCodeBundleID:(NSString*)codeBundleID {
-    [self bgSetAppValue:codeBundleID forKey:BSGKeyCodeBundleId];
+    [self setValue:codeBundleID forAppKey:BSGKeyCodeBundleId];
 }
 
-- (void)bgSetAppValue:(id)value forKey:(NSString*)key {
+- (void)setOrientation:(NSString *)orientation {
+    [self setValue:orientation forDeviceKey:BSGKeyOrientation];
+}
+
+- (void)setValue:(id)value forAppKey:(NSString *)key {
+    [self setValue:value forKey:key inDictionary:SYSTEMSTATE_KEY_APP];
+}
+
+- (void)setValue:(id)value forDeviceKey:(NSString *)key {
+    [self setValue:value forKey:key inDictionary:SYSTEMSTATE_KEY_DEVICE];
+}
+
+- (void)setValue:(id)value forKey:(NSString *)key inDictionary:(NSString *)dictionary {
     // Run on a BG thread so we don't monopolize the notification queue.
     dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
         @synchronized (self) {
-            self.currentLaunchStateRW[SYSTEMSTATE_KEY_APP][key] = value;
+            self.currentLaunchStateRW[dictionary][key] = value;
```

Comment:
The name `dictionary` will be confusing to someone not familiar with the code, and they might wonder if the type is wrong. Maybe some name that shows the hierarchy of the structure being modified (key/subkey maybe?).